src/sentry/static/sentry/app/views/dashboardsV2/dashboard.tsx

```diff
@@ -89,34 +143,228 @@ class Dashboard extends React.Component<Props, State> {
+      return;
+    }
+
+    event.stopPropagation();
+
+    // prevent the user from selecting things when dragging a widget.
+    this.previousUserSelect = setBodyUserSelect({
+      userSelect: 'none',
+      MozUserSelect: 'none',
+      msUserSelect: 'none',
+      webkitUserSelect: 'none',
+    });
+
+    // attach event listeners so that the mouse cursor can drag anywhere
+    window.addEventListener('mousemove', this.onWidgetDragMove);
+    window.addEventListener('mouseup', this.onWidgetDragEnd);
+
+    const widgetWrappers = document.querySelectorAll<HTMLDivElement>(
+      '[data-component="widget-wrapper"]'
+    );
+
+    const widgetWrapper = widgetWrappers[index];
+
+    const rects = widgetWrapper.getBoundingClientRect();
+
+    if (this.dragGhostRef.current) {
+      const ghostDOM = this.dragGhostRef.current;
+
+      // create the ghost widget
+
+      const newClone = widgetWrapper.cloneNode(true) as HTMLDivElement;
+      newClone.removeAttribute('data-component');
+
+      const iconContainer = newClone.querySelector('[data-component="icon-container"]');
+      if (iconContainer && iconContainer.parentNode) {
+        iconContainer.parentNode.removeChild(iconContainer);
+      }
+
+      newClone.style.width = `${rects.width}px`;
+      newClone.style.height = `${rects.height}px`;
+
+      ghostDOM.appendChild(newClone);
+
+      ghostDOM.style.left = `${event.pageX - GHOST_WIDGET_OFFSET}px`;
+      ghostDOM.style.top = `${event.pageY - GHOST_WIDGET_OFFSET}px`;
+    }
+
+    this.setState({
+      isDragging: true,
+      draggingIndex: index,
+      draggingTargetIndex: index,
+      top: event.pageY,
+      left: event.pageX,
+      widgets: this.shallowCloneWidgets(),
+    });
+  };
+
+  onWidgetDragMove = (event: MouseEvent) => {
+    if (
+      !this.state.isDragging ||
+      event.type !== 'mousemove' ||
+      !this.state.widgets ||
+      this.state.draggingIndex === undefined
+    ) {
+      return;
+    }
+
+    if (this.dragGhostRef.current) {
+      // move the ghost box
+      const ghostDOM = this.dragGhostRef.current;
+      ghostDOM.style.left = `${event.pageX - GHOST_WIDGET_OFFSET}px`;
+      ghostDOM.style.top = `${event.pageY - GHOST_WIDGET_OFFSET}px`;
+    }
+
+    const widgetWrappers = document.querySelectorAll<HTMLDivElement>(
+      '[data-component="widget-wrapper"]'
+    );
+
+    // Find the widget that the ghost is currently over.
+    const targetIndex = Array.from(widgetWrappers).findIndex(widgetWrapper => {
+      const rects = widgetWrapper.getBoundingClientRect();
+      const top = event.clientY;
+      const left = event.clientX;
+
+      const topStart = rects.top;
+      const topEnd = rects.top + rects.height;
+
+      const leftStart = rects.left;
+      const leftEnd = rects.left + rects.width;
+
+      return topStart <= top && top <= topEnd && leftStart <= left && left <= leftEnd;
+    });
+
+    if (targetIndex >= 0 && targetIndex !== this.state.draggingTargetIndex) {
```

Comment:
Do you also need to handle if the ghost is dragged to the very end of the list, or over the add placeholder?

Reply:
> Do you also need to handle if the ghost is dragged to the very end of the list, 

If the ghost is dragged to the very end of the list, it'll be reordered to the end of the list. Unsure if there's an edge case that was missed. 

Elements matching the selector `[data-component="widget-wrapper"]` are used for the hit detection relative to the ghost widget. 

> or over the add placeholder?

Dragging the ghost over the add placeholder doesn't do anything. Do you expect the add placeholder to move?

Reply:
> If the ghost is dragged to the very end of the list, it'll be reordered to the end of the list. Unsure if there's an edge case that was missed.

I poked around with the UI locally and I couldn't find any positioning scenarios that were missed.

> Dragging the ghost over the add placeholder doesn't do anything. Do you expect the add placeholder to move?

You're right that doesn't make sense as the add button is placed after all the widgets.

When dragging, should the active widget also display in the dashboard? Most drag sorting UX that I have used inserts a placeholder or gap where the item being dragged would go so that there is space for the ghost the cursor is actively holding. Should we do something similar here?